src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java

```diff
@@ -5013,14 +5018,69 @@ protected JCTree methodDeclaratorRest(int pos,
                 if (token.kind == DEFAULT) {
                     accept(DEFAULT);
                     defaultValue = annotationValue();
+                    accept(SEMI);
                 } else {
                     defaultValue = null;
+                    accept(SEMI, tk -> Errors.Expected2(LBRACE, SEMI));
                 }
-                accept(SEMI);
                 if (token.pos <= endPosTable.errorEndPos) {
                     // error recovery
-                    skip(false, true, false, false);
+                    // look if there is a probable missing opening brace,
+                    // and if yes, parse as a block
+                    skip(false, true, !unclosedParameterList, !unclosedParameterList);
+                    boolean parseAsBlock;
                     if (token.kind == LBRACE) {
+                        parseAsBlock = true;
+                    } else if (unclosedParameterList) {
+                        parseAsBlock = false;
+                    } else {
+                        parseAsBlock = switch (token.kind) {
```

Comment:
can we move this "complex" recovery code in a separate method and put a comment in there?